### ISSUE TITLE
Several methods in web2.dav that should return deferreds do not

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1817]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#1817|
|Type|enhancement|
|Created|2006-06-14 03:54:39Z|

There are several cases in the DAV code where methods should be returning deferreds and are not.  This is apparent now that we have a request.locateResource() which always returns a deferred.  Time to clean house.

<details><summary>Searchable metadata</summary>

```
trac-id__1817 1817
type__enhancement enhancement
reporter__wsanchez wsanchez
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__dav dav
time__1150257279000000 1150257279000000
changetime__1150844127000000 1150844127000000
version__None None
owner__ 
cc__wsanchez cc__dreid
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

branch: mo-deferreds-1817

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

Worst ticket summary ever.

Well, probably not, but it is pretty terrible...

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

What is it that you need to know?  There's a crapload of methods that need to return deferreds.  You need a list?  Still working on it, hence the branch.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

Thanks, the new summary is fine.  It's now an actual verb phrase, and gives some useful information about the scope of the issue.  I could see this new summary in a list of issues and actually have a fair idea what this ticket is about.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

OK, fixed on mo-deferreds-1817.

Some bonus items on that branch:

* There are some updates to the property store mechanism on DAVResource.  There are there to minimize code churn in another branch, from which this branch is derived, when we merge the older branch forward.
* There are several fixes to bugs in the test suites and in some of the DAV code which became apparent due to the deferred changes.
* There are now tests for PROPPATCH, since http_PROPATCH was changed substantially here.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @wsanchez|
|-|-|

So many things happening.  This is going to be a pretty big review, but I'm going to make it smaller by fixing the coding standards stuff myself.

IDavResource.*Property:
 Should these methods actually raise the HTTPError instead of errbacking the returned deferred?  The implementation seems to suggest they don't in which case the docstrings are misleading.

DAVPropertyMixIn.listAllprop doesn't actually work:

```python
def doList(qnames):
    qnames = []
    for qname in qnames:
         <snip>
```

You can see how this might be a problem.  It should be fixed and a unittest added, because apparently there isn't one that actually tests listAllprop.  (or this should have been caught by it)

DAVPropertyMixIn.*deadProperty:
  These don't seem to be on the IDavResource interface.  Also the implementations don't return deferreds.  Surely it's possible that the underlying property store returns a deferred for these (just like it can for live properties.)

dav.util.davXMLFromStream:
  Shouldn't the stream is None be checked for in allDataFromStream?  Is it ever used anywhere else?

All the new imports in method.delete are unused.

method.mkcol.http_MKCOL:
```
#
# Check authentication and access controls
#
```

Nothing actually does what that comment says it's doing.

Nit-picking:
 1. I've never been a big fan of bare calls to defer.(succeed, fail)
 1. The usage of deferred generators still bothers me.   I actually find large chunks of code that use it harder to follow than normal deferred using code.
 1. ```import twisted.web2.dav.test.test_copy```

I probably missed some things, so after you address these issues submit it for review again.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @dreid|
|-|-|

```IDavResource.*Property```:

  Yes, I'm at a bit of a loss as to what the proper way to document errbacks is.  I think stuffing all that into the ```@result``` field is really clunky, and it's clearer to keep them is ```@raise```.  I sent an email to t-py about this, and when we figure out the right thing to do, I'll file another ticket and clean it up everywhere.

```DAVPropertyMixIn.listAllprop```:

  Yikes.  Fixed, and added an ```allprop``` (and ```propname```) test to the ```PROPFIND``` tests, so this is covered.  I also made sure that we test all live properties, instead of a subset, which in turn exercises all of the live property generation code.

```DAVPropertyMixIn.*deadProperty```:

  These methods are specific to the ```DAVResource``` implementation.  While all DAV resources have properties, I don't think that all DAV resources should necessarily have to implement a public interface to their dead property store, and so these methods don't belong in ```IDAVResource```.

  I'd rather defer deferring these methods to another branch, because we already have a rather large set of changes here, doing so is adding a new feature, rather than fixing a bug.  That is, the methods already fixed here have to be fixed in order to be correct, because they call into methods which (may) return a deferred.  The dead property store is not broken in its current implementation, because none of the dead stores are blocking today.

```dav.util.davXMLFromStream```:

  This function isn't modified for this branch, so a separate ticket would be appropriate.  But no, I do not think it is incorrect.  The reason that ```davXMLFromStream``` handles ```None``` specifically is because the XML parser will raise an exception if you hand it ```None```, which is not the behavior we want.  If ```allDataFromStream``` handled this, it wouldn't keep ```davXMLFromStream``` from handing ```None``` to the parser.

```method.mkcol.http_MKCOL```:

  Fixed.

Nits:

 1. That's cosmetic.  I don't see how defer is special from any other module.  And the bare calls are arguably faster.
 1. Tomayto, tomahto.  I used defgens when a lot of refactoring would have otherwise been necessary, and I find them a lot easier to read.  In the simpler cases, I did avoid them.
 1. Yes.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @wsanchez|
|-|-|

```
[FAIL]: twisted.web2.dav.test.test_prop.PROP.test_PROPPATCH_basic

Traceback (most recent call last):
  File "/Users/dreid/projects/Twisted/trunk/twisted/web2/dav/test/test_prop.py", line 234, in check_patch_xml
    "Incorrect status code for PROPPATCH: %s != %s"
twisted.trial.unittest.FailTest: Incorrect status code for PROPPATCH: 403 != 200
```

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @dreid|
|-|-|

Fixed; turns out xattr isn't on the build slaves.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @wsanchez|
|-|-|

Ok, go ahead.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

Resolved by r17255 (If you put a : between "Fixes" and the ticket number, trac doesn't recognize it as a tracker command).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

And then reverted at r17259 - win32 was failing a test.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

exarkun merged forward to mo-deferreds-1817-2

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `closed`|
|-|-|

(In [17309]) Merged mo-deferreds-1817-2.
Reviewed by dreid
Fixes [#1817](https://github.com/twisted/twisted/issues/1817)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
